File: Searching/AL Searching/Binary_Search_.py
# Recursive method of binary search


class Binary_Search:
    def __init__(self, array):
        self.arr = array
        self.len = len(array)

    def search(self, l, h, value):

        if l == h:
            if self.arr[l] == value:
                return l
            return "list element not found"
        else:
            global mid
            mid = (l+h) // 2

            if self.arr[mid] == value:
                return mid+1
            if value < self.arr[mid]:
                return self.search(l, mid, value)
            else:
                return self.search(mid+1, h, value)


if __name__ == "__main__":
    arr = [2, 5, 8, 12, 16, 23, 38, 56, 72, 91]
    bs = Binary_Search(arr)
    # bs.search(0, len(arr), 5)
    print(bs.search(0, len(arr), 0))

File: Searching/search/Binary_Search.cpp
#include <iostream>
using namespace std;

int binarySearch(int arr[], int n, int k)
{
    int s = 0;
    int e = n;
    while (s <= e)
    {
        int mid = (s + e) / 2;
        if (arr[mid] == k)
            return mid;
        else if (arr[mid] > k)
        {
            e = mid - 1;
        }
        else
            s = mid + 1;
    }
    return -1;
}

int main()
{
    int arr[] = {10, 20, 30, 40, 50};
    int n = sizeof(arr) / sizeof(arr[0]);
    cout << binarySearch(arr, n, 40) << endl;
    return 0;
}